File: BAEKJOON/java/14888_연산자_끼워넣기.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class Main {

    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static int n;
    public static int[] nums;
    public static int[] ops = new int[4];
    public static int maxNum = -1000000001;
    public static int minNum = 1000000001;

    public static void main(String[] args) throws IOException {
        input();
        solution();

        bw.flush();
        bw.close();
    }

    private static void input() throws IOException {
        n = Integer.parseInt(br.readLine());

        nums = new int[n];

        String[] str = br.readLine().split(" ");
        for (int i = 0; i < n; i++) {
            nums[i] = Integer.parseInt(str[i]);
        }

        str = br.readLine().split(" ");
        for (int i = 0; i < 4; i++) {
            ops[i] = Integer.parseInt(str[i]);
        }
    }

    private static void solution() throws IOException {
        List<Character> operators = getOperators();

        boolean[] visit = new boolean[operators.size()];
        List<Character> list = new ArrayList<>();

        dfs(0, operators, visit, list);

        bw.write(String.valueOf(maxNum) + "\n" + String.valueOf(minNum));
    }

    private static void dfs(int cnt, List<Character> operators, boolean[] visit, List<Character> list) {
        if (cnt == operators.size()) {
            maxNum = Math.max(maxNum, calculate(list));
            minNum = Math.min(minNum, calculate(list));
        } else {
            for (int i = 0; i < operators.size(); i++) {
                if (visit[i] == true) continue;
                visit[i] = true;
                list.add(operators.get(i));
                dfs(cnt + 1, operators, visit, list);
                visit[i] = false;
                list.remove(list.size() - 1);
            }
        }
    }

    private static int calculate(List<Character> list) {
        int result = nums[0];

        for (int i = 0; i < list.size(); i++) {
            if (list.get(i) == '+') {
                result += nums[i + 1];
            } else if (list.get(i) == '-') {
                result -= nums[i + 1];
            } else if (list.get(i) == '*') {
                result *= nums[i + 1];
            } else {
                if (result < 0) {
                    result *= -1;
                    result /= nums[i + 1];
                    result *= -1;
                } else {
                    result /= nums[i + 1];
                }
            }
        }

        return result;
    }

    private static List<Character> getOperators() {
        List<Character> ret = new ArrayList<>();

        for (int i = 0; i < ops[0]; i++) {
            ret.add('+');
        }

        for (int i = 0; i < ops[1]; i++) {
            ret.add('-');
        }

        for (int i = 0; i < ops[2]; i++) {
            ret.add('*');
        }

        for (int i = 0; i < ops[3]; i++) {
            ret.add('/');
        }

        return ret;
    }
}

File: BAEKJOON/cpp/14888_연산자_끼워넣기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int n;
vector<int> nums;
vector<int> ops;
int maxNum = -1000000001;
int minNum = 1000000000;

void input() {
	cin >> n;

	int a;

	for (int i = 0; i < n; i++) {
		cin >> a;
		nums.push_back(a);
	}

	for (int i = 0; i < 4; i++) {
		cin >> a;
		ops.push_back(a);
	}
}

vector<char> getOperators() {
	vector<char> ret;

	for (int i = 0; i < ops[0]; i++) {
		ret.push_back('+');
	}

	for (int i = 0; i < ops[1]; i++) {
		ret.push_back('-');
	}

	for (int i = 0; i < ops[2]; i++) {
		ret.push_back('*');
	}

	for (int i = 0; i < ops[3]; i++) {
		ret.push_back('/');
	}

	return ret;
}

int calculate(vector<char>& v) {
	int result = nums[0];

	for (int i = 0; i < v.size(); i++) {
		if (v[i] == '+') {
			result += nums[i + 1];
		}
		else if (v[i] == '-') {
			result -= nums[i + 1];
		}
		else if (v[i] == '*') {
			result *= nums[i + 1];
		}
		else {
			if (result < 0) {
				result *= -1;
				result /= nums[i + 1];
				result *= -1;
			}
			else {
				result /= nums[i + 1];
			}
		}
	}

	return result;
}

void dfs(int cnt, vector<char>& operators, vector<bool>& visit, vector<char>& v) {
	if (cnt == operators.size()) {
		maxNum = max(maxNum, calculate(v));
		minNum = min(minNum, calculate(v));
	}
	else {
		for (int i = 0; i < operators.size(); i++) {
			if (visit[i] == true) continue;
			visit[i] = true;
			v.push_back(operators[i]);
			dfs(cnt + 1, operators, visit, v);
			visit[i] = false;
			v.pop_back();
		}
	}
}

void solution() {
	vector<char> operators = getOperators();

	vector<bool> visit(operators.size(), false);
	vector<char> v;

	dfs(0, operators, visit, v);

	cout << maxNum << "\n" << minNum;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	input();
	solution();

	return 0;
}
